### PR TITLE
Build bottles on Travis-CI and deploy to Bintray

### PR DESCRIPTION
**Pull request by davidchall**
The aim is to use Travis-CI to build bottles for this tap's formulae, and then upload them to my [Bintray](https://bintray.com/davidchall/homebrew-hep).

@MikeMcQuaid seems keen for this functionality in https://github.com/Homebrew/brew/issues/2572 and https://github.com/cartr/qt4-autobottler/issues/1. This would be helpful for other tap maintainers, and might be incorporated into [`brew tap-new`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/dev-cmd/tap-new.rb).

I believe I can do this using `brew test-bot` and `brew test-bot --ci-upload` if the right environment variables are set and arguments are passed. Some changes to [`brew test-bot`](https://github.com/Homebrew/homebrew-test-bot/blob/master/cmd/brew-test-bot.rb) might be required.

Notes:
- I set `HOMEBREW_BINTRAY_KEY` using encrypted Travis-CI variables

### CODE REVIEW

**Comment by davidchall** [comment]
There appears to be a conflict because `GIT_DIR` is set to the tap repo, which causes `brew update` to fail: https://github.com/Homebrew/homebrew-test-bot/blob/master/cmd/brew-test-bot.rb#L1202

```
$ brew test-bot --ci-upload --git-name=davidchall --git-email=dhcrawley@gmail.com --bintray_org=davidchall
Homebrew/homebrew-test-bot e43a85a (Merge pull request #122 from MikeMcQuaid/fix-linux)
ARGV: --ci-upload --git-name=davidchall --git-email=dhcrawley@gmail.com --bintray_org=davidchall --verbose --ci-auto --no-pull --coverage --ci-pr --cleanup --test-default-formula
git checkout -f master
Warning: you are leaving 2 commits behind, not connected to
any of your branches:
  6e1354f Merge e741771942638706731f5647200cfd59c90f59a9 into 730bdaba3d1ddcd3b3f158d6c3323181c3b160c0
  e741771 Initial attempt to build bottles
If you want to keep them by creating a new branch, this may be a good time
to do so with:
 git branch <new-branch-name> 6e1354f
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
git reset --hard origin/master
HEAD is now at 730bdab fjcontrib 1.030 (#110)
brew update
fatal: $program_name cannot be used without a working tree.
Error: Could not 'git stash' in /usr/local/Homebrew!
Please stash/commit manually if you need to keep your changes or, if not, run:
  cd /usr/local/Homebrew
  git reset --hard origin/master
Error: Failure while executing: brew update
```

**Comment by MikeMcQuaid** [comment]
This seems like a good idea, thanks @davidchall. The main thing you may want to additionally check is to only upload on PRs from non-forks. This unfortunately is also the possible blocker for this functionality being widely useful but it still seems like something useful to have in `brew tap-new`.

**Comment by davidchall** [comment]
@MikeMcQuaid Could you please elaborate on why we should only upload on PRs from non-forks?

Since taps often do get PRs from forks, I was thinking that we should do the upload after a PR has been accepted. That is, upload from non-PR builds on the master branch.

**Comment by MikeMcQuaid** [comment]
> Could you please elaborate on why we should only upload on PRs from non-forks?

Unfortunately Travis CI only exposes "secret" environment variables to PRs from non-forks i.e. the Bintray key needed for upload.

**Comment by davidchall** [comment]
>> Could you please elaborate on why we should only upload on PRs from non-forks?

> Unfortunately Travis CI only exposes "secret" environment variables to PRs from non-forks i.e. the Bintray key needed for upload.

Right, of course. That's why I thought uploading from master branch builds (after PR merge) might be better. Do you see any problem with this approach?

**Comment by MikeMcQuaid** [comment]
@davidchall The main problem is you'll need to push a commit somehow with the updated bottle checksums. Broadly that does seem like a good approach, though. It's also something that could work pretty well for single-maintainer taps where you're the main person creating the PRs.

**Comment by davidchall** [comment]
@MikeMcQuaid These changes, in combination with my [changes to `homebrew-test-bot`](https://github.com/Homebrew/homebrew-test-bot/compare/master...davidchall:master) now upload bottles to my Bintray repo from PRs, using the same package structure as Homebrew.

I'm struggling to edit `brew-test-bot` to produce bottles on master branch builds, though. This is pretty different from your current workflow of having an independent BrewTestBot handle this, which is based on PRs. Right now, it doesn't seem to notice that formulae have changed, so nothing is built. Would you be able to provide some advice on this step?

Once this barrier is overcome, `brew test-bot --ci-upload` already commits the bottle DSL changes, and it's possible to `git push` this to the master branch. Of course, this assumes that there are no conflicts in the formulae involved - but for small taps like the ones we're thinking about, it is probably fine.

File: .travis.yml [hunk 1]
```diff
@@ -24,10 +24,22 @@ before_install:
   - rm -rf "$HOMEBREW_TAP_DIR"
   - ln -s "$PWD" "$HOMEBREW_TAP_DIR"
   - export HOMEBREW_DEVELOPER="1"
+  - export HOMEBREW_BINTRAY_USER="davidchall"
+  - export TAP_BOTTLE_DOMAIN="https://dl.bintray.com/davidchall"
   - ulimit -n 1024
 
 script:
-  - brew test-bot
+  - brew tap davidchall/test-bot
```
Comment (MikeMcQuaid) [inline .travis.yml]: Would love you to submit a PR for some of these changes upstream 😁

File: .travis.yml [hunk 2]
```diff
@@ -24,10 +24,22 @@ before_install:
...
+  - if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then
+      brew test-bot --ci-testing;
+    else
+      brew test-bot --ci-pr;
+    fi
+
+after_success:
+  - if [ "$TRAVIS_BRANCH" = "master" ]; then
```
Comment (MikeMcQuaid) [inline .travis.yml]: Could maybe just do/detect this on any non-fork-PR branch?

File: .travis.yml (hunk 1 above)
Comment (davidchall) [reply .travis.yml]: I'll make sure to open a PR sometime soon.

File: .travis.yml (hunk 2 above)
Comment (davidchall) [reply .travis.yml]: When to upload the bottle is a matter of workflow. I think we should discuss this when I open a PR on the test-bot repo. For now, I just want a proof of concept.

**Comment by maelvalais** [comment]
Hi @davidchall,

First of all, thank you for all the (invaluable) information in this PR. I was looking for a way of using `test-bot` on a tap with travis+bintray.

That would be great if we could summarize the whole thing (things like: when testing travis builds, you must force a commit to trigger a "real" test-bot build). Too bad the test-bot project isn't that documented 😞 

Anyway, thanks for your work!!

**Comment by maelvalais** [comment]
If I understand correctly the workflow:
- to create a new release of some formulae, create a PR
- the `test-bot --ci-pr` will commit the `bottle` lines corresponding to the binaries into the PR.
  But at that time, no upload yet.
- we rebase the multiple commits made by travis into one commit
- when the PR is accepted and is merged, TravisCI will build and upload using `--ci-upload`; this is the moment where the binaries are uploaded to Bintray

Did I understand correctly?

**Comment by MikeMcQuaid** [comment]
> If you don't change the formula, brew test-bot will not build the bottle and everything.

Yep.

> Too bad the test-bot project isn't that documented 😞

We'd welcome PRs to improve that!

> during the PR TravisCI build, test-bot --ci-pr will push 3 commits corresponding to the binaries hashes into the PR:

`test-bot --ci-pr` will only commit onto the local Git repository in the Travis CI instance. You'll need to upload the resulting JSON files somewhere and combine them for the final commit to be done in the `--ci-upload` step.

**Comment by maelvalais** [comment]
Thank you so much!
So what I did was to set up a staged CI using TravisCI (travis is [here](https://travis-ci.org/touist/homebrew-touist), github for homebrew-touist is [here](https://github.com/touist/homebrew-touist)):
capture d ecran 2017-11-08 a 19 51 10

1. First stage is running `brew test-bot` only; at the end, they upload the bottle+json to an AWS S3 bucket (free).
2. Second stage ("deploy") only runs on master; it download the bottles+json from AWS S3 and uploads the bottles to Bintray and creates and pushes the commit with the merged bottle DSL stuff.

I forked `homebrew-test-bot` [here](https://github.com/maelvalais/homebrew-test-bot); I added some comments (for `--help`) and I enabled the git push using `https` instead of ssh.

There still is a problem with the rebuilds needed **in case of updated dependencies** as @scpeters mentionned in https://github.com/Homebrew/brew/issues/2572#issuecomment-298725510 and also mentionned in https://github.com/Homebrew/brew/issues/3346. I only rely on `gmp` but still; I should find a way to subscribe to the updates of gmp and trigger a revision+rebuild on my TravisCI instance.

**Comment by scpeters** [comment]
Nice work

**Comment by MikeMcQuaid** [comment]
> I forked homebrew-test-bot here; I added some comments (for --help) and I enabled the git push using https instead of ssh.

@maelvalais Could you submit a pull request for your changes?

**Comment by MikeMcQuaid** [comment]
> So what I did was to set up a staged CI using TravisCI (.travis.yml here, travis builds here, tap homebrew-touist is here):

Similarly a PR to Homebrew/brew's `tap-new.rb` would be great for Travis CI, too!

**Comment by maelvalais** [comment]
Here is the PR for `test-bot`: https://github.com/davidchall/homebrew-test-bot/pull/1
I'll take a look to [`tap-new`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/dev-cmd/tap-new.rb) and draft a PR.

**Comment by davidchall** [comment]
Hi @maelvalais, thanks for all your hard work on this. I'm in the middle of a coast-to-coast move, so things have slowed down a bit. But I'm more than willing to open a PR against `Homebrew/homebrew-test-bot` if this is what people desire.

I was keeping notes on the steps I took to set this up, with the idea to update [How-to-Create-and-Maintain-a-Tap](https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap.html) once I had the entire workflow ready. But it seems that you reverse-engineered a lot of this already.

I originally just focussed on creating and uploading the bottles to Bintray, and was then planning to figure out how to use staged Travis builds to push the commit that adds the bottle DSL to the formula. Looks like you've got this figured out, and in a smart way. Thank you!!

**Comment by MikeMcQuaid** [comment]
@davidchall @maelvalais Sorry, I was meaning could the two of you submit your changes back up to the main homebrew-test-bot and Homebrew/brew repos. No rush, though.

**Comment by maelvalais** [comment]
Thanks @davidchall :)

I began to write some notes for How-to-Create-and-Maintain-a-Tap in a [gist](https://gist.github.com/maelvalais/068af21911c7debc4655cdaa41bbf092); I wrote them in case I forget something.

**Comment by MikeMcQuaid** [comment]
@davidchall Any updates here? No worries if not. Thanks!

**Comment by maelvalais** [comment]
Hi!
I merged (like 1h ago) the current Linuxbrew's test-bot into @davidchall's fork. I'll open a PR for that; some conflicts were kind of tricky and I had to `git blame` a lot in order to find  what was what. I hope we will manage to get it merged 😇

**Comment by davidchall** [comment]
Hi @MikeMcQuaid @maelvalais, I plan to get back into this project over the next couple of weeks and review the latest PR. Sounds like we're getting very close to merging back into the Homebrew repo.

**Pull request merged**